src/engine/Match_tainting_mode.ml

```diff
@@ -132,6 +132,30 @@ type propagator_match = {
 }
 (** Taint will flow from `from` to `to_` through the axiliary variable `id`. *)
 
+(* This Formula_tbl structure is used to create a "formula cache", which will
+   permit the sharing of matches resulting from common formulas that occur as
+   sources, sinks, sanitizers, or propagators.
+
+   In particular, because of hardcoded propagators, we expect to see lots of
+   sharing from Semgrep Pro Engine.
+*)
+module Formula_tbl = struct
+  include Hashtbl.Make (struct
+    type t = Rule.formula
+
+    let equal = AST_utils.with_structural_equal Rule.equal_formula
+    let hash = Rule.hash_formula
+  end)
+
+  let cached_find_opt formula_cache formula cont =
```

Comment:
I'm not sure about passing this `cont` here... but let's check the rest of the code first.

Reply:
In any case I think `cont` is a bad name, when I see `cont` I expect to see CPS or something, so you would apply `cont` to the result of lookup up the formula in the cache. But this argument is a _delayed_ `range_w_metas_of_formula`.

Reply:
sure, thats fair. changed!